src/app/util/attribute-storage.cpp

```diff
@@ -266,9 +267,82 @@ uint16_t emberAfGetDynamicIndexFromEndpoint(EndpointId id)
     return kEmberInvalidEndpointIndex;
 }
 
+namespace {
+
+const EmberAfCluster * getClusterTypeDefinition(EndpointId endpointId, ClusterId clusterId, EmberAfClusterMask mask)
+{
+    uint16_t index = emberAfIndexFromEndpointIncludingDisabledEndpoints(endpointId);
+    if (index != kEmberInvalidEndpointIndex)
+    {
+        return emberAfFindClusterInType(emAfEndpoints[index].endpointType, clusterId, mask, nullptr);
+    }
+    // not found
+    return nullptr;
+}
+
+} // anonymous namespace
+
+CHIP_ERROR emberAfSetupDynamicEndpointDeclaration(EmberAfEndpointType & endpointType, EndpointId templateEndpointId,
+                                                  const Span<const EmberAfClusterSpec> & templateClusterSpecs)
+{
+    // we want an explicitly empty endpoint to begin with, to make sure no already set-up endpoint is passed in
+    VerifyOrReturnError(endpointType.cluster == nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(endpointType.endpointSize == 0, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(endpointType.clusterCount == 0, CHIP_ERROR_INVALID_ARGUMENT);
+    // check cluster count fits target struct's clusterCount field
+    size_t clusterCount = templateClusterSpecs.size();
+    VerifyOrReturnError(CanCastTo<typeof(endpointType.clusterCount)>(clusterCount), CHIP_ERROR_NO_MEMORY);
+    // allocate new cluster array
+    auto newClusters = new EmberAfCluster[clusterCount];
+    VerifyOrReturnError(newClusters != nullptr, CHIP_ERROR_NO_MEMORY);
+    size_t endpointSize = 0;
+    // get the actual cluster pointers and sum up memory size
+    for (size_t i = 0; i < clusterCount; i++)
+    {
+        auto cluster =
+            getClusterTypeDefinition(templateEndpointId, templateClusterSpecs[i].clusterId, templateClusterSpecs[i].mask);
+        if (!cluster)
+        {
+            delete[] newClusters;
+            ChipLogError(DataManagement, "cluster 0x%04x with mask %x could not be found in template endpoint %u",
+                         (unsigned int) templateClusterSpecs[i].clusterId, templateClusterSpecs[i].mask,
+                         (unsigned int) templateEndpointId);
+            return CHIP_ERROR_NOT_FOUND;
+        }
+        // for now, we need to copy the cluster definition, unfortunately.
+        // TODO: make endpointType use a pointer to a list of EmberAfCluster* instead, so we can re-use cluster definitions
+        //   instead of duplicating them here once for every instance.
+        newClusters[i] = *cluster;
+        // sum up the needed storage, result must
+        endpointSize += cluster->clusterSize;
+        if (!CanCastTo<typeof(endpointType.endpointSize)>(endpointSize))
```

Comment:
```suggestion
        if (!CanCastTo<decltype(endpointType.endpointSize)>(endpointSize))
```

Typeof seems to be C++23. We support C++17 and up .https://en.cppreference.com/w/c/language/typeof